## **Notebook 12 — RAG + Long/Short Term Memory (End to End)**
Integrates the full production RAG pipeline with a persistent memory system.

- Document: data/docs/md_al_amin_resume.pdf
- STM: LangGraph PostgresSaver checkpointer — per-thread conversation history, summarised at 6 messages
- LTM: LangGraph PostgresStore — persistent user profile facts across all threads
- The LLM answers from the resume AND from what it remembers about the user across sessions.

# ***PART 1 — Imports & Model Loading***

## **Cell 1 — install + imports**
- **Why:** All libraries needed for the full pipeline in one place.
- **langgraph-checkpoint-postgres:** provides PostgresSaver (STM) and PostgresStore (LTM). These are official LangGraph persistence backends — they store conversation checkpoints and memory namespaces directly in your existing PostgreSQL database. No separate vector database needed for memory.


In [1]:
# Install if needed
# !pip install langgraph-checkpoint-postgres psycopg[binary] psycopg[pool]
# !pip install FlagEmbedding bm25s docling langchain-openai python-dotenv

import os, json, re, uuid, operator
from pathlib import Path
from typing import TypedDict, Annotated, Optional
from dataclasses import dataclass, field as dc_field
from collections import Counter
from datetime import datetime
from dotenv import load_dotenv

import numpy as np
import psycopg2
from pgvector.psycopg2 import register_vector
import bm25s

from docling.document_converter import DocumentConverter
from FlagEmbedding import BGEM3FlagModel, FlagReranker
from langchain_openai import ChatOpenAI
from langchain_core.messages import (
    SystemMessage, HumanMessage, AIMessage, RemoveMessage
)

# LangGraph core
from langgraph.graph import StateGraph, END
from langgraph.checkpoint.postgres import PostgresSaver
from langgraph.store.postgres import PostgresStore

load_dotenv()

DATABASE_SYNC = os.getenv("DATABASE_URL_SYNC")
OPENAI_KEY    = os.getenv("OPENAI_API_KEY")
INDEX_DIR     = Path("../data/bm25_index_resume")
INDEX_DIR.mkdir(parents=True, exist_ok=True)

print("Imports OK")
print(f"DB  : {DATABASE_SYNC[:10]}...")
print(f"Key : {OPENAI_KEY[:15]}...")

Imports OK
DB  : postgresql...
Key : sk-proj-Pm8wgjk...


## **Cell 2 — load AI models**
- **Why load all models here:** BGE-M3 (~2.3GB) and the reranker (~1.1GB) take 20-30s to load. Loading once at the top means every subsequent cell is instant.
- **GPT-4o-mini for everything:** We use gpt-4o-mini for HyDE generation, LTM fact extraction, STM summarisation, and final answer generation. It is fast, cheap, and accurate enough for all four tasks.

In [2]:
MODEL_DIR = Path("../models")

print("Loading BGE-M3 embedder...")
embed_model = BGEM3FlagModel(
    "BAAI/bge-m3", use_fp16=False,
    cache_dir=str(MODEL_DIR / "bge-m3")
)
print("  BGE-M3 ready")

print("Loading BGE reranker...")
rerank_model = FlagReranker(
    "BAAI/bge-reranker-v2-m3", use_fp16=False,
    cache_dir=str(MODEL_DIR / "bge-reranker")
)
print("  Reranker ready")

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0.1)
print("  LLM ready (gpt-4o-mini)")

print("\nAll models loaded OK")

Loading BGE-M3 embedder...


Fetching 30 files:   0%|          | 0/30 [00:00<?, ?it/s]

  BGE-M3 ready
Loading BGE reranker...
  Reranker ready
  LLM ready (gpt-4o-mini)

All models loaded OK


## ***Cell 3 — fresh DB tables for this notebook**
- **Why fresh tables:** Isolated from previous notebooks so you can re-run from scratch without corrupting earlier experiments.
chunks_resume + documents_resume: Same schema as before but with suffix to avoid collisions.
- **pgvector HNSW index:** Created immediately after table creation since we have few chunks (resume is small). In production with large corpora, create the HNSW index AFTER bulk loading all data for performance.

In [3]:
conn_str = DATABASE_SYNC.replace("postgresql+psycopg2://", "postgresql://")
conn     = psycopg2.connect(conn_str)
register_vector(conn)
cur      = conn.cursor()

# Drop and recreate for clean start
cur.execute("DROP TABLE IF EXISTS chunks_resume   CASCADE;")
cur.execute("DROP TABLE IF EXISTS documents_resume CASCADE;")

cur.execute("""
    CREATE TABLE documents_resume (
        id         UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        filename   TEXT NOT NULL,
        status     TEXT DEFAULT 'ready',
        created_at TIMESTAMPTZ DEFAULT now()
    );
""")

cur.execute("""
    CREATE TABLE chunks_resume (
        id             UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        document_id    UUID REFERENCES documents_resume(id) ON DELETE CASCADE,
        content        TEXT NOT NULL,
        parent_content TEXT,
        embedding      vector(1024),
        metadata       JSONB NOT NULL DEFAULT '{}',
        created_at     TIMESTAMPTZ DEFAULT now()
    );
""")

cur.execute("""
    CREATE INDEX idx_resume_embedding
    ON chunks_resume
    USING hnsw (embedding vector_cosine_ops)
    WITH (m=16, ef_construction=64);
""")

cur.execute("""
    CREATE INDEX idx_resume_metadata
    ON chunks_resume USING gin(metadata);
""")

conn.commit()
print("Tables created: documents_resume, chunks_resume")
print("Indexes created: HNSW (embedding), GIN (metadata)")

Tables created: documents_resume, chunks_resume
Indexes created: HNSW (embedding), GIN (metadata)


## **Cell 4 — parse resume PDF with Docling**
- **Why Docling over PyPDF2:** Your resume has structured sections — Education, Experience, Projects, Skills. Docling preserves these as markdown headers which become section metadata on every chunk. PyPDF2 would flatten them into one block of text.
- **Path:**  data/docs/md_al_amin_resume.pdf — the file you uploaded.

In [4]:
PDF_PATH = Path("../data/docs/md_al_amin_resume.pdf")

if not PDF_PATH.exists():
    print(f"PDF not found at {PDF_PATH}")
    print("Make sure md_al_amin_resume.pdf is in data/docs/")
else:
    print(f"Found: {PDF_PATH}")
    print(f"Size : {PDF_PATH.stat().st_size / 1024:.1f} KB")

converter = DocumentConverter()
print("\nParsing with Docling...")
result   = converter.convert(str(PDF_PATH))
markdown = result.document.export_to_markdown()

print(f"Parsed successfully")
print(f"Characters: {len(markdown)}")
print(f"\nMarkdown preview:")
print(markdown[:600])

Found: ../data/docs/md_al_amin_resume.pdf
Size : 117.2 KB

Parsing with Docling...


[INFO] 2026-05-06 11:32:02,910 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-06 11:32:02,921 [RapidOCR] download_file.py:60: File exists and is valid: /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-05-06 11:32:02,922 [RapidOCR] main.py:53: Using /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-05-06 11:32:02,982 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-06 11:32:02,984 [RapidOCR] download_file.py:60: File exists and is valid: /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer.onnx
[INFO] 2026-05-06 11:32:02,984 [RapidOCR] main.py:53: Using /home/md-al-amin/My-Projects/End-to-End-Agentic-Ai-Automation-Lab/.venv/lib/python3.12/site

Parsed successfully
Characters: 3766

Markdown preview:
## Education

## North South University

Bachelor of Science in Computer Science and Engineering

Jan 18 - Apr 25

CGPA: 3.17 (86%)

|

Coursework: Data Structures, Algorithms, DBMS, ML, DL, NLP, Operating Systems

## Technical Skills

Languages : Python | C/C++ | SQL | JavaScript

GenAI &amp; ML : PyTorch | TensorFlow | LLM | LangChain | LangGraph | vLLM | MCP | Prompt and Context Engineering | Embedding | RAG | Multi-Model RAG | PEFT(LoRA, QLoRA) | VectorDB | OpenAI API | Whisper | STT Web &amp; DevOps : FastAPI, Docker, GitLab CI/CD, Redis, NGiNX, PostgreSQL, RunPod, Playwright, Microservic


## **Cell 5 — chunking functions**
- **Why smaller chunks for a resume:** A resume has dense, structured content — each bullet point is a complete fact. We use child_size=2 sentences and parent_size=4 sentences. Larger windows would blend Skills with Experience which hurts retrieval precision.
- **Section detection:** The detect_section function walks backwards through the markdown lines to find the nearest ## header. Every chunk gets tagged with its section — "Experience", "Projects", "Skills" etc. This enables filtered queries like "only search within Skills section".

In [5]:
def detect_section(lines: list[str], up_to: int) -> str:
    """Walk backwards from current position to find nearest ## header."""
    for line in reversed(lines[:up_to]):
        line = line.strip()
        if line.startswith("## "):
            return line.lstrip("#").strip()
        if line.startswith("# "):
            return line.lstrip("#").strip()
    return "General"

def split_sentences(text: str) -> list[str]:
    """Split into sentences, skip headers and short fragments."""
    sentences = []
    for line in text.split("\n"):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        parts = re.split(r'(?<=[.!?])\s+', line)
        for part in parts:
            if len(part.strip()) > 15:
                sentences.append(part.strip())
    return sentences

def build_chunks(markdown: str, filename: str,
                 child_size: int = 2,
                 parent_size: int = 4,
                 overlap: int = 1) -> list[dict]:
    """
    Build parent-child chunks from markdown text.

    child_size=2  — 2 sentences per child (precise retrieval embedding)
    parent_size=4 — 4 sentences per parent (full context for LLM)
    overlap=1     — 1 sentence shared between adjacent chunks
                    prevents answers that span chunk boundaries from being missed

    Returns list of dicts with content, parent_content, section, metadata.
    """
    sentences = split_sentences(markdown)
    lines     = markdown.split("\n")
    chunks    = []
    idx       = 0
    i         = 0

    while i < len(sentences):
        parent_sents   = sentences[i : i + parent_size]
        parent_content = " ".join(parent_sents)
        section        = detect_section(lines, min(i * 2, len(lines) - 1))

        j = 0
        while j < len(parent_sents):
            child_content = " ".join(parent_sents[j : j + child_size])
            if len(child_content.strip()) > 15:
                chunks.append({
                    "content"       : child_content,
                    "parent_content": parent_content,
                    "section"       : section,
                    "metadata"      : {
                        "filename"   : filename,
                        "section"    : section,
                        "chunk_index": idx,
                    }
                })
                idx += 1
            j += child_size - overlap
        i += parent_size - overlap

    return chunks

chunks = build_chunks(markdown, "md_al_amin_resume.pdf")
print(f"Total chunks: {len(chunks)}")
print(f"\nSection distribution:")
for section, count in Counter(c["section"] for c in chunks).most_common():
    print(f"  {section:<35} {count:>3} chunks")
print(f"\nSample chunk:")
print(f"  Child  : {chunks[0]['content'][:100]}")
print(f"  Parent : {chunks[0]['parent_content'][:100]}")

Total chunks: 41

Section distribution:
  North South University                8 chunks
  Genuine Technology &amp; Research Ltd   8 chunks
  Jr. Generative AI Engineer            8 chunks
  1. SynapseAI High-Performance Memory Agent | FastAPI, LangGraph, Redis, PostgreSQL   8 chunks
  General                               4 chunks
  Technical Skills                      4 chunks
  3. Agentic AI Math Tutor with Adaptive Prompting | LangGraph, ChromaDB, Llama-4 [Capstone Proj-Apr 2025]   1 chunks

Sample chunk:
  Child  : Bachelor of Science in Computer Science and Engineering CGPA: 3.17 (86%)
  Parent : Bachelor of Science in Computer Science and Engineering CGPA: 3.17 (86%) Coursework: Data Structures


## **Cell 6 — embed chunks + store in pgvector**
- **Why batch_size=16:** Resume has ~38 chunks. A single batch is fine. For larger documents use batch_size=32 and loop.
- **What gets stored:** The child chunk content, parent chunk content, the 1024-dim BGE-M3 dense embedding, and JSONB metadata. The embedding is what pgvector searches. The parent content is what the LLM receives.

In [6]:
texts    = [c["content"] for c in chunks]
print(f"Embedding {len(texts)} chunks with BGE-M3...")

output   = embed_model.encode(
    texts,
    batch_size=16,
    return_dense=True,
    return_sparse=False,
)
vecs = output["dense_vecs"]
print(f"Embeddings shape: {vecs.shape}")

# Insert document record
cur.execute(
    "INSERT INTO documents_resume (filename) VALUES (%s) RETURNING id;",
    ("md_al_amin_resume.pdf",)
)
doc_id = cur.fetchone()[0]

# Insert all chunks
for chunk, vec in zip(chunks, vecs):
    cur.execute("""
        INSERT INTO chunks_resume
            (document_id, content, parent_content, embedding, metadata)
        VALUES (%s, %s, %s, %s, %s::jsonb)
    """, (
        doc_id,
        chunk["content"],
        chunk["parent_content"],
        vec.tolist(),
        json.dumps(chunk["metadata"]),
    ))

conn.commit()

cur.execute("SELECT COUNT(*) FROM chunks_resume;")
print(f"Stored {cur.fetchone()[0]} chunks in pgvector")

Embedding 41 chunks with BGE-M3...


pre tokenize: 100%|██████████| 3/3 [00:00<00:00, 47.57it/s]
You're using a XLMRobertaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
Inference Embeddings: 100%|██████████| 3/3 [00:00<00:00,  6.11it/s]


Embeddings shape: (41, 1024)
Stored 41 chunks in pgvector


## **Cell 7 — build BM25s index**

- **Why BM25s on a resume:** Resumes are full of exact terms — "LangGraph", "FastAPI", "CGPA 3.17", "North South University", "July 2025". Dense embeddings may miss these exact strings because they represent rare proper nouns. BM25s catches them reliably.
- **Load chunk_ids from DB:** We fetch IDs in insertion order so the BM25s index array position maps exactly to the correct chunk UUID in pgvector.

In [7]:
chunk_texts = [c["content"] for c in chunks]

cur.execute("SELECT id::text FROM chunks_resume ORDER BY created_at;")
chunk_ids = [r[0] for r in cur.fetchall()]

print(f"Building BM25s index over {len(chunk_texts)} chunks...")
corpus_tokens  = bm25s.tokenize(chunk_texts, stopwords="en")
bm25_retriever = bm25s.BM25()
bm25_retriever.index(corpus_tokens)

bm25_retriever.save(str(INDEX_DIR / "bm25_index"))
with open(INDEX_DIR / "chunk_ids.json",   "w") as f: json.dump(chunk_ids,   f)
with open(INDEX_DIR / "chunk_texts.json", "w") as f: json.dump(chunk_texts, f)

print(f"BM25s index saved — {len(chunk_ids)} entries")
print("Ingestion complete")

Building BM25s index over 41 chunks...


Split strings:   0%|          | 0/41 [00:00<?, ?it/s]

BM25S Count Tokens:   0%|          | 0/41 [00:00<?, ?it/s]

BM25S Compute Scores:   0%|          | 0/41 [00:00<?, ?it/s]

BM25s index saved — 41 entries
Ingestion complete


A# ***RT 3 — Memory System Setup***


## **Cell 8 — setup PostgresSaver (STM) and PostgresStore (LTM)**

- **PostgresSaver (STM):** Stores the conversation message history for each thread. When you pass a thread_id config to the graph, LangGraph automatically saves and restores the message list from this checkpointer. Each thread has its own independent history.
- **PostgresStore (LTM):** Stores arbitrary key-value facts in namespaces. We use namespace ("user", user_id, "profile") for user profile facts. These persist across ALL threads for that user — unlike the checkpointer which is thread-scoped.
- **Connection string format:** PostgresSaver and PostgresStore require the psycopg3 connection string format — postgresql://user:pass@host:port/db.

In [8]:
from psycopg_pool import ConnectionPool
from langgraph.checkpoint.postgres import PostgresSaver
from langgraph.store.postgres import PostgresStore
from functools import partial

# Use the same DB URL pattern from your old notebook
# Your .env has DATABASE_URL_SYNC — convert to psycopg3 format
PG_CONN_STR = DATABASE_SYNC.replace("postgresql+psycopg2://", "postgresql://")

# ── Connection pool — shared by both STM and LTM ─────────────
# max_size=20 — max concurrent connections in the pool
# autocommit=True — required by both PostgresSaver and PostgresStore
pool = ConnectionPool(
    conninfo=PG_CONN_STR,
    max_size=20,
    kwargs={"autocommit": True},
)

# ── STM: PostgresSaver ────────────────────────────────────────
# Pass the pool directly — not from_conn_string
checkpointer = PostgresSaver(pool)
checkpointer.setup()   # creates langgraph_checkpoints table
print("STM checkpointer (PostgresSaver) ready")

# ── LTM: PostgresStore ────────────────────────────────────────
# Same pool — no second connection needed
ltm_store = PostgresStore(pool)
ltm_store.setup()      # creates langgraph_store table
print("LTM store (PostgresStore) ready")

print()
print("Memory tables in PostgreSQL:")
print("  langgraph_checkpoints — STM (conversation history per thread)")
print("  langgraph_store       — LTM (user profile facts across threads)")

STM checkpointer (PostgresSaver) ready
LTM store (PostgresStore) ready

Memory tables in PostgreSQL:
  langgraph_checkpoints — STM (conversation history per thread)
  langgraph_store       — LTM (user profile facts across threads)


## **Cell 9 — MemoryOp schema + LTM manager prompt**

- **Why MemoryOp schema:** Instead of asking the LLM to write free-form memory updates, we define a strict JSON schema with three operations: create, update, delete. The LLM must respond with a list of these operations. This makes memory management deterministic and auditable — you can see exactly what facts were added, changed, or removed after every message.
- **The LTM manager prompt:** After every user message + assistant response pair, we send both to the LLM with this prompt. It decides which facts (if any) to create, update, or delete in the user's profile. If nothing important was said, it returns an empty list.

In [9]:
from pydantic import BaseModel

class MemoryOp(BaseModel):
    """
    One memory operation — create, update, or delete a fact about the user.

    action : "create"  — store a new fact that was not known before
             "update"  — correct or extend an existing fact
             "delete"  — remove a fact that is no longer true

    key    : short identifier for the fact, e.g. "name", "role", "university"
    value  : the fact content (empty string for delete operations)
    reason : brief explanation of why this operation is needed
    """
    action : str    # "create" | "update" | "delete"
    key    : str    # e.g. "name", "role", "location", "goal"
    value  : str    # the fact value
    reason : str    # why this operation

LTM_MANAGER_PROMPT = """You are a memory manager for an AI assistant.

Your job: after reading the conversation exchange below, decide what facts
about the USER should be stored, updated, or deleted in their long-term profile.

Rules:
- Only store facts the user explicitly stated about themselves
- Do NOT store opinions, questions, or temporary states
- Do NOT store facts from the document (resume) — only facts the user told you
- If the user corrects a previous statement, update the old fact
- If nothing important was said, return an empty list
- Keep fact values short and factual (under 30 words)

User message : {user_message}
Assistant response: {assistant_response}

Return a JSON array of memory operations. Each must have:
  action : "create" | "update" | "delete"
  key    : short snake_case identifier
  value  : the fact (empty for delete)
  reason : why this operation

Return [] if nothing should be stored. Return only valid JSON, no other text."""

STM_SUMMARY_PROMPT = """Summarise this conversation in 3-5 sentences.
Focus on: what the user asked about, what topics were covered, any important
context that would help answer future questions in this conversation.
Be concise. Do not include greetings or filler.

Conversation:
{messages}

Summary:"""

print("MemoryOp schema defined")
print("LTM_MANAGER_PROMPT defined")
print("STM_SUMMARY_PROMPT defined")

MemoryOp schema defined
LTM_MANAGER_PROMPT defined
STM_SUMMARY_PROMPT defined


## **Cell 10 — LTM helper functions**

- **get_ltm_profile:** Loads all stored facts for a user from PostgresStore. Returns a formatted string like "name: Al Amin | role: AI Engineer | goal: PhD scholarship". This string is injected into the generation prompt so the LLM knows who it is talking to.
- **apply_memory_ops:** Executes a list of MemoryOp objects against PostgresStore. Create and Update both call store.put() — pgstore is upsert-by-default. Delete calls store.delete(). This function is called by the update_memory node after every response.


In [10]:
def get_ltm_profile(store, user_id: str) -> str:
    """
    Load all LTM facts for a user and return as a formatted string.

    Queries PostgresStore namespace ("user", user_id, "profile").
    Returns formatted string for injection into generation prompt.
    Returns empty string if no facts exist yet (new user).

    Example output:
        "name: Al Amin | role: Jr. AI Engineer | goal: PhD scholarship"
    """
    namespace = ("user", user_id, "profile")
    try:
        items = store.search(namespace)
        if not items:
            return ""
        facts = [f"{item.key}: {item.value}" for item in items
                 if isinstance(item.value, str)]
        return " | ".join(facts)
    except Exception:
        return ""

def apply_memory_ops(store, user_id: str, ops: list[MemoryOp]) -> int:
    """
    Execute a list of MemoryOp objects against PostgresStore.

    create / update → store.put(namespace, key, value)
    delete          → store.delete(namespace, key)

    Returns count of operations applied.
    """
    namespace = ("user", user_id, "profile")
    applied   = 0

    for op in ops:
        try:
            if op.action in ("create", "update"):
                store.put(namespace, op.key, op.value)
                applied += 1
            elif op.action == "delete":
                store.delete(namespace, op.key)
                applied += 1
        except Exception as e:
            print(f"  [memory] op failed: {op.action} {op.key} — {e}")

    return applied

def extract_memory_ops(user_message: str, assistant_response: str) -> list[MemoryOp]:
    """
    Ask gpt-4o-mini to extract memory operations from a conversation turn.
    Returns list of MemoryOp objects (may be empty if nothing to store).
    """
    prompt   = LTM_MANAGER_PROMPT.format(
        user_message       = user_message,
        assistant_response = assistant_response,
    )
    response = llm.invoke([HumanMessage(content=prompt)])
    raw      = response.content.strip()

    # Strip markdown code fences if present
    if raw.startswith("```"):
        raw = re.sub(r"^```[a-z]*\n?", "", raw)
        raw = re.sub(r"\n?```$",        "", raw)

    try:
        ops_data = json.loads(raw)
        return [MemoryOp(**op) for op in ops_data if isinstance(op, dict)]
    except Exception:
        return []

print("LTM helper functions defined:")
print("  get_ltm_profile()      — loads user profile from PostgresStore")
print("  apply_memory_ops()     — executes create/update/delete ops")
print("  extract_memory_ops()   — LLM extracts ops from conversation turn")

LTM helper functions defined:
  get_ltm_profile()      — loads user profile from PostgresStore
  apply_memory_ops()     — executes create/update/delete ops
  extract_memory_ops()   — LLM extracts ops from conversation turn


## **Cell 11 — STM helper functions**

- **get_stm_summary:** Reads the current conversation summary from the LangGraph checkpointer state. LangGraph stores messages as a list in the checkpoint. We look for a special summary message we injected previously.

- **summarise_and_trim:** When message count exceeds 6, we: (1) ask gpt-4o-mini to summarise the full conversation, (2) generate RemoveMessage objects for all old messages, (3) return the summary text + remove operations. The LangGraph state graph applies the removes automatically via the operator.add annotation on the messages field, which handles RemoveMessage correctly.

In [11]:
STM_WINDOW  = 6    # trigger summarisation when message count exceeds this
STM_KEEP    = 2    # keep this many recent messages after summarisation

def get_messages_from_state(state: dict) -> list:
    """Extract the messages list from the LangGraph state."""
    return state.get("messages", [])

def should_summarise(messages: list) -> bool:
    """Return True if message count exceeds STM_WINDOW."""
    return len(messages) > STM_WINDOW

def summarise_messages(messages: list) -> str:
    """
    Ask gpt-4o-mini to summarise the conversation history.
    Called when message count exceeds STM_WINDOW.

    Returns a plain text summary string.
    """
    # Format messages for the prompt
    formatted = []
    for msg in messages:
        if isinstance(msg, HumanMessage):
            formatted.append(f"User: {msg.content}")
        elif isinstance(msg, AIMessage):
            formatted.append(f"Assistant: {msg.content}")
    conversation_text = "\n".join(formatted)

    prompt   = STM_SUMMARY_PROMPT.format(messages=conversation_text)
    response = llm.invoke([HumanMessage(content=prompt)])
    return response.content.strip()

def build_trim_ops(messages: list, keep_last: int = STM_KEEP) -> list:
    """
    Build a list of RemoveMessage operations to trim old messages.
    Keeps the last `keep_last` messages.
    Returns list of RemoveMessage objects.

    LangGraph handles RemoveMessage via the operator.add annotation:
    when the state update contains RemoveMessage, LangGraph removes
    that message from the list instead of appending.
    """
    if len(messages) <= keep_last:
        return []
    messages_to_remove = messages[:-keep_last]
    return [RemoveMessage(id=m.id) for m in messages_to_remove
            if hasattr(m, "id") and m.id]

print("STM helper functions defined:")
print("  should_summarise()   — checks if message count > STM_WINDOW")
print("  summarise_messages() — LLM summarises conversation history")
print("  build_trim_ops()     — builds RemoveMessage list for trimming")

STM helper functions defined:
  should_summarise()   — checks if message count > STM_WINDOW
  summarise_messages() — LLM summarises conversation history
  build_trim_ops()     — builds RemoveMessage list for trimming


## **Cell 12 — define RAGMemState**

### **RAGMemState extends the RAGState from Notebook 11 with three memory fields:**
- `user_id` — scopes all LTM operations to one user. Every `store.put()` and `store.search()` uses namespace ("user", user_id, "profile").
- `ltm_context` — the formatted user profile string loaded by load_memory. Injected into the generation prompt alongside retrieved chunks.
- `stm_summary` — the rolling conversation summary. Injected into the generation prompt so the LLM knows what was discussed earlier in this thread.
- `messages` — the full message list managed by the LangGraph checkpointer. Uses `Annotated[list, operator.add]` so both append and RemoveMessage operations work correctly.

In [12]:
class RAGMemState(TypedDict):
    """
    Full state object for the RAG + Memory pipeline.
    Flows through every node — each node reads what it needs
    and returns only the fields it changed.

    Memory fields (new in this notebook):
        user_id     — identifies the user for LTM scoping
        ltm_context — user profile from PostgresStore (loaded by load_memory)
        stm_summary — conversation summary from checkpointer (loaded by load_memory)
        messages    — full message list (managed by LangGraph checkpointer)

    RAG fields (same as Notebook 11):
        query, hyde_doc, hyde_embedding,
        retrieved_chunks, fused_chunks, reranked_chunks,
        answer, sources, confidence, retry_count, error
    """
    # ── Identity ──────────────────────────────────────────────
    user_id       : str     # e.g. "user_001" — scopes LTM namespace

    # ── Memory ────────────────────────────────────────────────
    ltm_context   : str     # "name: Al Amin | role: AI Engineer | ..."
    stm_summary   : str     # rolling conversation summary

    # Message list — operator.add enables both append AND RemoveMessage
    messages      : Annotated[list, operator.add]

    # ── RAG ───────────────────────────────────────────────────
    query          : str
    hyde_doc       : Optional[str]
    hyde_embedding : Optional[list]
    retrieved_chunks: Annotated[list, operator.add]
    fused_chunks   : list
    reranked_chunks: list
    answer         : Optional[str]
    sources        : list
    confidence     : float
    retry_count    : int
    error          : Optional[str]

print("RAGMemState defined")
print("Fields:", list(RAGMemState.__annotations__.keys()))

RAGMemState defined
Fields: ['user_id', 'ltm_context', 'stm_summary', 'messages', 'query', 'hyde_doc', 'hyde_embedding', 'retrieved_chunks', 'fused_chunks', 'reranked_chunks', 'answer', 'sources', 'confidence', 'retry_count', 'error']


## **Cell 14 — nodes 2–6: RAG retrieval nodes**

- **These are identical to Notebook 11 with one change:** transform_query receives ltm_context in its HyDE prompt. If the user's profile says they are an AI engineer looking for PhD programs, the hypothetical answer will use more relevant vocabulary for their queries.
- **All retrieval nodes are self-contained** — they read from pgvector and BM25s, not from memory. Memory only affects the query transformation and generation steps.

In [13]:
HYDE_PROMPT = """You are a resume and career expert.
Write a short passage (3-5 sentences) that directly answers the question below.
Write as if the passage was extracted from a professional resume or career profile.
Use formal language. Do NOT say "I". Just write the passage directly.

{memory_context}
Question: {query}
Passage:"""

def transform_query(state: RAGMemState) -> dict:
    """
    NODE 2 — HyDE Query Transformation

    Uses LTM context to generate a more targeted hypothetical document.
    If user profile says 'role: AI Engineer', the HyDE doc will use
    AI/ML vocabulary that better matches resume content.
    """
    print(f"  [transform_query] Query: '{state['query'][:60]}'")
    try:
        mem_hint = ""
        if state.get("ltm_context"):
            mem_hint = f"User context: {state['ltm_context']}\n"

        prompt   = HYDE_PROMPT.format(
            memory_context=mem_hint, query=state["query"]
        )
        response = llm.invoke([HumanMessage(content=prompt)])
        hyde_doc = response.content
        hyde_vec = embed_model.encode(
            [hyde_doc], return_dense=True
        )["dense_vecs"][0].tolist()

        print(f"  [transform_query] HyDE: '{hyde_doc[:70]}'")
        return {"hyde_doc": hyde_doc, "hyde_embedding": hyde_vec}

    except Exception as e:
        raw_vec = embed_model.encode(
            [state["query"]], return_dense=True
        )["dense_vecs"][0].tolist()
        return {"hyde_doc": state["query"], "hyde_embedding": raw_vec,
                "error": f"HyDE fallback: {e}"}

def retrieve_dense(state: RAGMemState) -> dict:
    """NODE 3 — Dense retrieval from pgvector using HyDE embedding."""
    print(f"  [retrieve_dense] Searching pgvector...")
    try:
        query_vec = state["hyde_embedding"]
        cur.execute("""
            SELECT id::text, content, parent_content,
                   metadata->>'section',
                   1-(embedding <=> %s::vector) AS sim
            FROM chunks_resume
            ORDER BY embedding <=> %s::vector
            LIMIT 20;
        """, (query_vec, query_vec))
        results = [
            {"chunk_id": r[0], "content": r[1], "parent_content": r[2],
             "section": r[3], "score": float(r[4]),
             "rank": i+1, "source": "dense"}
            for i, r in enumerate(cur.fetchall())
        ]
        print(f"  [retrieve_dense] Found {len(results)} chunks")
        return {"retrieved_chunks": results}
    except Exception as e:
        print(f"  [retrieve_dense] ERROR: {e}")
        return {"retrieved_chunks": [], "error": str(e)}

def retrieve_sparse(state: RAGMemState) -> dict:
    """NODE 4 — Sparse retrieval from BM25s index."""
    print(f"  [retrieve_sparse] BM25s search...")
    try:
        tokens = bm25s.tokenize([state["query"]], stopwords="en")
        results, scores = bm25_retriever.retrieve(
            tokens, k=min(20, len(chunk_ids))
        )
        sparse = [
            {"chunk_id": chunk_ids[idx], "content": chunk_texts[idx],
             "parent_content": None, "section": None,
             "score": float(s), "rank": i+1, "source": "sparse"}
            for i, (idx, s) in enumerate(zip(results[0], scores[0]))
        ]
        print(f"  [retrieve_sparse] Found {len(sparse)} chunks")
        return {"retrieved_chunks": sparse}
    except Exception as e:
        print(f"  [retrieve_sparse] ERROR: {e}")
        return {"retrieved_chunks": [], "error": str(e)}

def fuse_results(state: RAGMemState) -> dict:
    """NODE 5 — RRF fusion of dense + sparse results."""
    print(f"  [fuse_results] Fusing {len(state['retrieved_chunks'])} candidates...")
    all_c  = state["retrieved_chunks"]
    dense  = [c for c in all_c if c.get("source") == "dense"]
    sparse = [c for c in all_c if c.get("source") == "sparse"]

    scores = {}
    for r in dense:
        scores[r["chunk_id"]] = scores.get(r["chunk_id"],0)+1/(60+r["rank"])
    for r in sparse:
        scores[r["chunk_id"]] = scores.get(r["chunk_id"],0)+1/(60+r["rank"])

    ranked   = sorted(scores.items(), key=lambda x:x[1], reverse=True)[:10]
    d_lookup = {r["chunk_id"]:r for r in dense}
    fused    = []
    for cid, rrf in ranked:
        base = d_lookup.get(cid, {})
        fused.append({
            "chunk_id"       : cid,
            "content"        : base.get("content",""),
            "parent_content" : base.get("parent_content"),
            "section"        : base.get("section"),
            "rrf_score"      : rrf,
            "rank"           : len(fused)+1,
            "in_dense"       : cid in {r["chunk_id"] for r in dense},
            "in_sparse"      : cid in {r["chunk_id"] for r in sparse},
        })

    # Enrich sparse-only chunks with parent_content from DB
    missing = [r for r in fused if not r.get("parent_content")]
    if missing:
        ph  = ",".join(["%s"]*len(missing))
        cur.execute(
            f"SELECT id::text, content, parent_content, metadata->>'section' "
            f"FROM chunks_resume WHERE id::text IN ({ph});",
            [r["chunk_id"] for r in missing]
        )
        enriched = {r[0]: r for r in cur.fetchall()}
        for r in fused:
            if r["chunk_id"] in enriched:
                e = enriched[r["chunk_id"]]
                r["content"]        = r["content"]        or e[1]
                r["parent_content"] = r["parent_content"] or e[2]
                r["section"]        = r["section"]        or e[3]

    print(f"  [fuse_results] Fused to {len(fused)} candidates")
    return {"fused_chunks": fused}

def rerank_chunks(state: RAGMemState) -> dict:
    """NODE 6 — Cross-encoder reranking. Sets confidence score."""
    print(f"  [rerank_chunks] Reranking {len(state['fused_chunks'])} candidates...")
    candidates = state["fused_chunks"]
    if not candidates:
        return {"reranked_chunks": [], "confidence": 0.0}

    pairs  = [[state["query"], c["content"]] for c in candidates]
    scores = rerank_model.compute_score(pairs, normalize=True)
    for c, s in zip(candidates, scores): c["rerank_score"] = float(s)
    reranked = sorted(candidates, key=lambda x:x["rerank_score"], reverse=True)
    for i, r in enumerate(reranked): r["rank"] = i+1
    top5 = reranked[:5]
    confidence = float(np.mean([r["rerank_score"] for r in reranked[:3]])) \
                 if reranked else 0.0

    # print(f"  [rerank_chunks] Scores: {[f'{r[\"rerank_score\"]:.3f}' for r in top5]}")
    scores = [f"{r['rerank_score']:.3f}" for r in top5]
    print(f"  [rerank_chunks] Scores: {scores}")
    print(f"  [rerank_chunks] Confidence: {confidence:.3f}")
    return {"reranked_chunks": top5, "confidence": confidence}

print("RAG retrieval nodes defined (2-6)")

RAG retrieval nodes defined (2-6)


## **Cell 15 — node 7: generate_answer**

**Three-section prompt: This is the key difference from Notebook 11. The generation prompt now has three context sections:**

1. Document context [1][2][3] — retrieved resume chunks (what the document says)
2. User profile (LTM) — facts the user told us across all sessions
3. Conversation history (STM) — summary of what was discussed in this thread

*The LLM uses all three to produce a personalised, contextually aware answer. If the user told us they are applying for a PhD in Canada and asks "what skills do I have for NLP research?", the answer frames the resume skills specifically for a Canadian PhD application — because the LTM tells the bot that goal.*

In [14]:
SYSTEM_PROMPT_WITH_MEMORY = """You are a helpful career and resume assistant.

Answer questions using:
1. The provided document passages (cite with [1][2] etc.)
2. What you know about the user from their profile
3. The conversation history context

If the document passages don't contain the answer, say so clearly.
Never invent facts not present in the passages or user profile.
Use the user's profile to personalise your answer when relevant."""

def generate_answer(state: RAGMemState) -> dict:
    """
    NODE 7 — LLM Generation with Memory Context

    Three-section prompt:
    1. Retrieved document chunks (resume content)
    2. LTM user profile (what user told us across sessions)
    3. STM summary (what was discussed in this thread)

    The LLM sees all three and generates a personalised,
    grounded answer with citations.

    Also appends the new (user_query, answer) pair to the
    messages list for the STM checkpointer to persist.
    """
    print(f"  [generate_answer] Generating answer...")
    chunks = state["reranked_chunks"]

    # 1. Build numbered document context
    doc_parts = []
    for i, chunk in enumerate(chunks, 1):
        text    = chunk.get("parent_content") or chunk.get("content", "")
        section = chunk.get("section", "Unknown")
        doc_parts.append(f"[{i}] (Section: {section})\n{text}")
    doc_context = "\n\n".join(doc_parts) if doc_parts else "No relevant passages found."

    # 2. LTM user profile section
    ltm_section = ""
    if state.get("ltm_context"):
        ltm_section = f"\n\nUSER PROFILE (what you know about this user):\n{state['ltm_context']}"

    # 3. STM conversation summary section
    stm_section = ""
    if state.get("stm_summary"):
        stm_section = f"\n\nCONVERSATION HISTORY SUMMARY:\n{state['stm_summary']}"

    prompt = f"""DOCUMENT PASSAGES:
{doc_context}{ltm_section}{stm_section}

QUESTION: {state['query']}

ANSWER (cite document passages with [1][2] etc.):"""

    response = llm.invoke([
        SystemMessage(content=SYSTEM_PROMPT_WITH_MEMORY),
        HumanMessage(content=prompt),
    ])
    answer = response.content

    sources = [
        {
            "citation"     : f"[{i}]",
            "section"      : c.get("section", ""),
            "preview"      : (c.get("content") or "")[:80],
            "rerank_score" : c.get("rerank_score", 0),
        }
        for i, c in enumerate(chunks, 1)
    ]

    print(f"  [generate_answer] Answer: '{answer[:100]}'")

    # Append this exchange to the message list
    # The checkpointer persists these automatically
    new_messages = [
        HumanMessage(content=state["query"]),
        AIMessage(content=answer),
    ]

    return {
        "answer"  : answer,
        "sources" : sources,
        "messages": new_messages,
    }

print("generate_answer node defined")

generate_answer node defined


## **Cell 16 — node 8: update_memory**

**What this node does after every response:**

- **LTM update:** Sends the latest (user_query, answer) pair to the LTM manager prompt. The LLM decides if any facts about the user should be created, updated, or deleted in PostgresStore. This runs on every message — but usually returns an empty list unless the user revealed something about themselves.
- **STM summarisation:** Counts the messages in the checkpointer. If count exceeds STM_WINDOW=6, calls summarise_messages(), stores the summary as a SystemMessage, and returns RemoveMessage operations for all old messages. The next call to load_memory will find the summary and inject it into the prompt.
- **This is the fire-and-forget pattern:** In production you would run this as a background task after sending the response. Here in the notebook it runs synchronously to make the behaviour visible.

In [15]:
def update_memory(state: RAGMemState) -> dict:
    """
    NODE 8 — Update Memory (LTM + STM)

    Runs after every generate_answer call.

    LTM update:
        Extracts facts from the latest (query, answer) exchange.
        Applies create/update/delete operations to PostgresStore.

    STM management:
        If message count > STM_WINDOW (6):
          1. Summarise entire conversation with LLM
          2. Store summary as SystemMessage("Conversation summary: ...")
          3. Return RemoveMessage ops for all old messages
          4. Only the summary + last STM_KEEP (2) messages remain

    Returns:
        messages — list containing new SystemMessage(summary) + RemoveMessages
                   (empty list if no summarisation needed)
    """
    print(f"  [update_memory] Running memory update...")

    user_id  = state.get("user_id", "default_user")
    query    = state.get("query", "")
    answer   = state.get("answer", "")
    messages = state.get("messages", [])

    # ── LTM update ────────────────────────────────────────────
    ops = extract_memory_ops(query, answer)
    if ops:
        applied = apply_memory_ops(ltm_store, user_id, ops)
        print(f"  [update_memory] LTM: applied {applied} ops")
        for op in ops:
            print(f"    {op.action:8} '{op.key}' = '{op.value[:50]}'")
    else:
        print(f"  [update_memory] LTM: nothing to update")

    # ── STM management ────────────────────────────────────────
    if not should_summarise(messages):
        print(f"  [update_memory] STM: {len(messages)} messages, no summary needed")
        return {"messages": []}

    print(f"  [update_memory] STM: {len(messages)} messages — summarising...")

    # 1. Generate summary
    summary_text = summarise_messages(messages)
    print(f"  [update_memory] Summary: '{summary_text[:100]}'")

    # 2. Summary stored as a SystemMessage with a recognisable prefix
    summary_msg  = SystemMessage(content=f"Conversation summary: {summary_text}")

    # 3. RemoveMessage ops for all old messages (keep last STM_KEEP)
    remove_ops   = build_trim_ops(messages, keep_last=STM_KEEP)
    print(f"  [update_memory] STM: removing {len(remove_ops)} old messages")

    # Return: new summary message + remove ops
    # operator.add handles all of these correctly:
    #   - SystemMessage is appended normally
    #   - RemoveMessage causes that message to be deleted from the list
    return {"messages": [summary_msg] + remove_ops}

print("update_memory node defined")

update_memory node defined


## **Cell 17 — terminal nodes + conditional edge**

**Same conditional edge logic as Notebook 11** — confidence < 0.4 triggers retry, max 2 retries.
- `retry_retrieval` is simplified here — just re-runs retrieval with raw query embedding (no HyDE) and broader k=30. The memory state is preserved across retries.
- `Terminal` nodes append their refusal message to the messages list so the conversation history is complete — even failed queries are recorded.

In [16]:
def should_generate_or_retry(state: RAGMemState) -> str:
    """
    Conditional edge — routes after rerank_chunks.
    confidence >= 0.4 → generate_answer
    confidence <  0.4 AND retry < 2 → retry_retrieval
    confidence <  0.4 AND retry >= 2 → end_low_confidence
    """
    confidence  = state.get("confidence",  0.0)
    retry_count = state.get("retry_count", 0)
    error       = state.get("error")

    critical = ["database", "connection", "pgvector"]
    if error and any(e.lower() in str(error).lower() for e in critical):
        return "end_with_error"

    if confidence >= 0.4:
        return "generate"
    if retry_count < 2:
        return "retry"
    return "end_low_confidence"

def retry_retrieval(state: RAGMemState) -> dict:
    """Broader retrieval attempt on low confidence."""
    retry_num = state.get("retry_count", 0) + 1
    print(f"  [retry_retrieval] Retry #{retry_num}...")
    raw_vec = embed_model.encode(
        [state["query"]], return_dense=True
    )["dense_vecs"][0].tolist()
    cur.execute("""
        SELECT id::text, content, parent_content,
               metadata->>'section', 1-(embedding <=> %s::vector)
        FROM chunks_resume ORDER BY embedding <=> %s::vector LIMIT 30;
    """, (raw_vec, raw_vec))
    dense = [{"chunk_id":r[0],"content":r[1],"parent_content":r[2],
              "section":r[3],"score":float(r[4]),"rank":i+1,"source":"dense"}
             for i,r in enumerate(cur.fetchall())]
    tokens = bm25s.tokenize([state["query"]], stopwords="en")
    res, sc = bm25_retriever.retrieve(tokens, k=min(30,len(chunk_ids)))
    sparse  = [{"chunk_id":chunk_ids[idx],"content":chunk_texts[idx],
                "parent_content":None,"section":None,
                "score":float(s),"rank":i+1,"source":"sparse"}
               for i,(idx,s) in enumerate(zip(res[0],sc[0]))]
    scores = {}
    for r in dense+sparse:
        scores[r["chunk_id"]] = scores.get(r["chunk_id"],0)+1/(60+r["rank"])
    ranked   = sorted(scores.items(),key=lambda x:x[1],reverse=True)[:10]
    d_lookup = {r["chunk_id"]:r for r in dense}
    fused    = [{"chunk_id":cid,"content":d_lookup.get(cid,{}).get("content",""),
                 "parent_content":d_lookup.get(cid,{}).get("parent_content"),
                 "section":d_lookup.get(cid,{}).get("section"),
                 "rrf_score":rrf,"rank":i+1,"in_dense":cid in {r["chunk_id"] for r in dense},
                 "in_sparse":cid in {r["chunk_id"] for r in sparse}}
                for i,(cid,rrf) in enumerate(ranked)]
    pairs  = [[state["query"],c["content"]] for c in fused]
    sc_r   = rerank_model.compute_score(pairs, normalize=True)
    for c,s in zip(fused,sc_r): c["rerank_score"] = float(s)
    reranked = sorted(fused,key=lambda x:x["rerank_score"],reverse=True)[:5]
    for i,r in enumerate(reranked): r["rank"]=i+1
    confidence = float(np.mean([r["rerank_score"] for r in reranked[:3]])) if reranked else 0.0
    return {"retrieved_chunks":[],"fused_chunks":fused,
            "reranked_chunks":reranked,"confidence":confidence,"retry_count":retry_num}

def end_low_confidence(state: RAGMemState) -> dict:
    answer = (f"I could not find relevant information to answer: '{state['query']}'. "
              f"Please try rephrasing or check if the topic is in the document.")
    return {"answer":answer,"sources":[],
            "messages":[HumanMessage(content=state["query"]),AIMessage(content=answer)]}

def end_with_error(state: RAGMemState) -> dict:
    answer = f"Pipeline error: {state.get('error','Unknown error')}"
    return {"answer":answer,"sources":[],"messages":[]}

print("Terminal nodes + conditional edge defined")

Terminal nodes + conditional edge defined


## **Cell 18 — build + compile the graph**

- **Two new nodes vs Notebook 11:** load_memory at the start and update_memory at the end.
- **Checkpointer passed to compile():** This is the key line. When compile(checkpointer=checkpointer) is called, LangGraph automatically saves the full state (including the messages list) to PostgreSQL after every node. On the next invocation with the same thread_id, it restores the state from where it left off. This is how STM works — no manual serialisation needed.
- **Store passed to compile():** The store=ltm_store argument makes PostgresStore available to all nodes. Nodes access it via ltm_store (our module-level variable) directly here in the notebook. In the FastAPI version you would inject it via the LangGraph store API.